src/cryptography/hazmat/backends/openssl/x509.py

```diff
@@ -151,3 +152,62 @@ def signature_hash_algorithm(self):
             raise UnsupportedAlgorithm(
                 "Signature algorithm OID:{0} not recognized".format(oid)
             )
+
+    @property
+    def extensions(self):
+        # TODO: remove this when we finish landing support for extensions.
+        warnings.warn(
+            "Extension support is not fully implemented. Unsupported "
+            "extensions are IGNORED at this time."
+        )
+        extensions = []
+        seen_oids = set()
+        extcount = self._backend._lib.X509_get_ext_count(self._x509)
+        for i in range(0, extcount):
+            ext = self._backend._lib.X509_get_ext(self._x509, i)
+            assert ext != self._backend._ffi.NULL
+            crit = self._backend._lib.X509_EXTENSION_get_critical(ext)
+            critical = crit == 1
+            nid = self._backend._lib.OBJ_obj2nid(ext.object)
+            assert nid != self._backend._lib.NID_undef
+            if nid == self._backend._lib.NID_basic_constraints:
+                value = self._build_basic_constraints(ext)
+                oid = x509.OID_BASIC_CONSTRAINTS
+            else:
+                # Unsupported extension, silently skipping.
+                continue
+
+            # Check for duplicate extensions
+            if oid in seen_oids:
+                raise x509.DuplicateExtension(
+                    "Duplicate {0} extension found".format(oid)
+                )
+            else:
+                seen_oids.add(oid)
+
+            extensions.append(x509.Extension(oid, critical, value))
+
+        return x509.Extensions(extensions)
+
+    def _build_basic_constraints(self, ext):
+        bc_st = self._backend._lib.X509V3_EXT_d2i(ext)
+        assert bc_st != self._backend._ffi.NULL
+        bc = self._backend._ffi.cast("BASIC_CONSTRAINTS *", bc_st)
```

Comment:
Oof.

On Thu, Mar 26, 2015 at 8:09 PM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/hazmat/backends/openssl/x509.py
> <https://github.com/pyca/cryptography/pull/1769#discussion_r27268080>:
>
> > +            # Check for duplicate extensions
> > +            if oid in seen_oids:
> > +                raise x509.DuplicateExtension(
> > +                    "Duplicate {0} extension found".format(oid)
> > +                )
> > +            else:
> > +                seen_oids.add(oid)
> > +
> > +            extensions.append(x509.Extension(oid, critical, value))
> > +
> > +        return x509.Extensions(extensions)
> > +
> > +    def _build_basic_constraints(self, ext):
> > +        bc_st = self._backend._lib.X509V3_EXT_d2i(ext)
> > +        assert bc_st != self._backend._ffi.NULL
> > +        bc = self._backend._ffi.cast("BASIC_CONSTRAINTS *", bc_st)
>
> Yep...X509V3_EXT_d2i returns a void * which in reality corresponds to the
> internal object matching the extension type.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1769/files#r27268080>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084